deploy/helm/templates/statefulset.yaml

```
@@ -46,12 +46,24 @@ spec:
       initContainers:
       {{- if .Values.redis.enabled }}
       - name: redis-init-container
-        image: alpine
+      {{- if .Values.initContainer.redis }}
+      {{- if .Values.initContainer.redis.image }}
+        image: {{ .Values.initContainer.redis.image }}
+      {{- end }}
+      {{- else }}
+        image: "alpine"
+      {{- end }}
```

I see that in the `values.yaml`, we set the default for this as `initContainer: {}`, which is an empty map.

But here, we check that if the `redis` key is not there in that map, we don't set an `image:` at all. Am I reading this right? Can we change the templating so that we _always_ ensure that an `image:` field is set? Something like:

```suggestion
      {{- if and .Values.initContainer.redis .Values.initContainer.redis.image }}
        image: {{ .Values.initContainer.redis.image }}
      {{- else }}
        image: "alpine"
      {{- end }}
```

Similarly for the MongoDB container as well. What do you think?

---

In the above case the outer condition has the else block (line 53/57/59) and we set the default image in line 58, if the redis key doesn't exists. 

Unfortunately the suggested change with function "and" doesn't work here, see also [Helm optional nested variables](https://stackoverflow.com/questions/59795596/helm-optional-nested-variables)

Therefore I've added the possible implementation from SO above, see also latest commit.

Out of the box chart, no initContainers set:
```
$ helm template -n test appsmith-init-images . -s templates/statefulset.yaml | yq '.spec.template.spec.initContainers[].image'              
"alpine"                                                                     
"docker.io/bitnami/mongodb:4.4.11-debian-10-r12"
```

Setting initContainers in values.yaml:
```
$ helm template -n test appsmith-init-images . -s templates/statefulset.yaml --set redis.images=123456789012.dkr.ecr.eu-west-1.amazonaws.com:external/alpine --set mongodb.images=123456789012.dkr.ecr.eu-west-1.amazonaws.com:external/ubuntu | yq '.spec.template.spec.initContainers[].image'          
"123456789012.dkr.ecr.eu-west-1.amazonaws.com:external/alpine"
"123456789012.dkr.ecr.eu-west-1.amazonaws.com:external/ubuntu
```

---

Awesome, thanks for pointing me to that detail. I didn't know parentheses had that feature in Go's template module. TIL.

Also, I just noticed that I misrepresented the problem I was noticing in this (now outdated) version of your code. What I wanted to point out what that, if `redis` is set, but `redis.image` is not set, then we'll be missing the `image:` key. But along the way I got confused and talked about it wrongly. The parent `if` does indeed have an `else`, like you point out, it's the child `if` that's missing one. 🙂

But not a problem, I see you've resolved this now. I'll approve in a moment, but we'll merge this next week, is that okay? Just want to avoid merging in the holiday season if I can. 🙂

---

Thank you and sorry for the back and forth. :slightly_smiling_face: Therefore no worries with the merge, whenever you are comfortable.   :slightly_smiling_face:

Also, just noticed the above helm template test is not correct, just for completeness here the corrected values settings as parameter. 
```
$ helm template -n test appsmith-init-images . -s templates/statefulset.yaml --set initContainer.redis.image=123456789012.dkr.ecr.eu-west-1.amazonaws.com:external/alpine --set initContainer.mongodb.image=123456789012.dkr.ecr.eu-west-1.amazonaws.com:external/ubuntu | yq '.spec.template.spec.initContainers[].image'                                                                                                              
                                                                                                      
"123456789012.dkr.ecr.eu-west-1.amazonaws.com:external/alpine"
"123456789012.dkr.ecr.eu-west-1.amazonaws.com:external/ubuntu"
```

I'll close the other comments above, as they are outdated.